File: Common/Game Core/TFENode.m
//
//  TFENode.m
//  TwentyFortyEight
//
//  Created by Joshua Caswell on 2/28/16.
//  Copyright © 2016 Josh Caswell. All rights reserved.
//

#import "TFENode.h"

static CGFloat randomOffsetWithGamut(CGFloat gamut)
{
    return gamut * ((CGFloat)arc4random() / UINT32_MAX) - (gamut / 2);
}

@interface TFENode ()

+ (SKTexture *)textureForValue:(uint32_t)value;

/** The TFENode class tracks whether any instances are running animations.
 *  This dispatch group is the means. An instance enters the group before
 *  starting its animation action, and leaves the group in the action's
 *  completion.
 */
+ (dispatch_group_t)movementDispatchGroup;

- (void)moveToPosition:(CGPoint)destination duration:(CGFloat)duration;

@end

@implementation TFENode

+ (instancetype)nodeWithValue:(uint32_t)value
{
    SKTexture * texture = [self textureForValue:value];
    id node = [[self alloc] initWithTexture:texture];
    
    [node setSize:(CGSize){100, 100}];
    [(TFENode *)node setValue:value];
    
    return node;
}

+ (void)waitOnAllNodeMovement
{
    dispatch_group_wait([self movementDispatchGroup], DISPATCH_TIME_FOREVER);
}

+ (BOOL)anyNodeMovementInProgress
{
    // Return immediately regardless, but signal whether it was because
    // the group is inactive or because of timeout.
    return 0 != dispatch_group_wait([self movementDispatchGroup],
                                    DISPATCH_TIME_NOW);
}

+ (SKTexture *)textureForValue:(uint32_t)value
{
    NSString * name = [NSString stringWithFormat:@"%d", value];
    return [SKTexture textureWithImageNamed:name];
}

+ (dispatch_group_t)movementDispatchGroup
{
    static dispatch_group_t group;
    
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        group = dispatch_group_create();
    });
    
    return group;
}

- (BOOL)isEqual:(id)other
{
    if( ![other isKindOfClass:[self class]] ){
        return NO;
    }
    
    return [self value] == [(TFENode *)other value];
}

- (NSUInteger)hash
{
    return [[self class] hash] << [self value];
}

- (CGFloat)resizeFadeDuration
{
    static const CGFloat kNodeResizeFadeDuration = 0.19;
    CGFloat gamut = 0.08;
    CGFloat offset = randomOffsetWithGamut(gamut);
    return kNodeResizeFadeDuration + offset;
}

- (CGFloat)bounceDuration
{
    static const CGFloat kNodeBounceDuration = 0.06;
    CGFloat gamut = 0.08;
    CGFloat offset = randomOffsetWithGamut(gamut);
    return kNodeBounceDuration + offset;
}

- (CGFloat)moveDuration
{
    static const CGFloat kNodeMoveDuration = 0.125;
    CGFloat gamut = 0.08;
    CGFloat offset = randomOffsetWithGamut(gamut);
    return kNodeMoveDuration + offset;
}

-(void)spawnAtPosition:(CGPoint)position
{
    static const CGFloat kInitialScale = 0.85;
    static const CGFloat kBounceUpperScale = 1.1;
    static const CGFloat kBounceLowerScale = 0.93;
    static const CGFloat kUnityScale = 1.0;

    CGFloat thisResizeDuration = [self resizeFadeDuration];
    CGFloat thisBounceDuration = [self bounceDuration];
    
    // Grow to full size
    [self setScale:kInitialScale];
    SKAction * grow = [SKAction scaleTo:kUnityScale duration:thisResizeDuration];
    
    // Do a little size bounce
    SKAction * scaleUp = [SKAction scaleTo:kBounceUpperScale duration:thisBounceDuration];
    SKAction * scaleDown = [SKAction scaleTo:kBounceLowerScale duration:thisBounceDuration];
    SKAction * scaleBackUp = [SKAction scaleTo:kUnityScale duration:thisBounceDuration];
    SKAction * bounceScale = [SKAction sequence:@[scaleUp, scaleDown, scaleBackUp]];
    
    SKAction * spawn = [SKAction sequence:@[grow, bounceScale]];

    [self setPosition:position];
    [self runAction:spawn];
}

- (void)moveToPosition:(CGPoint)destination
{
    [self moveToPosition:destination duration:[self moveDuration]];
}

- (void)moveIntoCombinationAtPosition:(CGPoint)destination
{
    static const CGFloat kFinalShrinkScale = 0.25;
    
    BOOL changingPosition = !CGPointEqualToPoint(destination, [self position]);
    
    CGFloat thisResizeDuration = [self resizeFadeDuration];
    SKAction * shrink = [SKAction scaleTo:kFinalShrinkScale duration:thisResizeDuration];
    SKAction * fade = [SKAction fadeOutWithDuration:thisResizeDuration];
    
    SKAction * fadeAndShrink = [SKAction group:@[shrink, fade]];

    if( changingPosition ){

        CGFloat thisMoveDuration = [self moveDuration];
        fadeAndShrink = [SKAction sequence:@[[SKAction waitForDuration:thisMoveDuration], fadeAndShrink]];
        [self moveToPosition:destination duration:thisMoveDuration];
    }
    
    // Disappear animation does not prevent spawning
    [self runAction:fadeAndShrink
         completion:^{
        
            // Next run loop
            dispatch_async(dispatch_get_main_queue(), ^{
                [self removeFromParent];
            });
    }];
}

- (void)moveToPosition:(CGPoint)destination duration:(CGFloat)duration
{
    SKAction * move = [SKAction moveTo:destination duration:duration];
    [move setTimingMode:SKActionTimingEaseInEaseOut];
    
    dispatch_group_enter([TFENode movementDispatchGroup]);
    [self runAction:move
         completion:^{ dispatch_group_leave([TFENode movementDispatchGroup]); }];
}

@end

File: Common/Game Core/TFENode.swift
//
//  TFENode.swift
//  TwentyFortyEight
//
//  Created by Joshua Caswell on 10/3/16.
//  Copyright © 2016 Josh Caswell. All rights reserved.
//

import SpriteKit

/**
 * TFENodes are the sprites representing the tiles on the game board. They configure their appearance
 * and perform necessary animations. They also hold their point value, which the game engine uses to
 * calculate combinations.
 */
class TFENode : SKSpriteNode
{
    /** Create the appropriate `SKTexture` for this value. */
    private static func texture(forValue value: Int) -> SKTexture
    {
        return SKTexture(imageNamed: "\(value)")
    }
    
    /** The node's point value. This is a `TFENode`'s sole feature for purposes of comparison to other nodes. */
    let value: Int
    
    init(value: Int)
    {
        self.value = value
        
        let texture = TFENode.texture(forValue: value)
        super.init(texture: texture, color: SKColor.clear, size: CGSize(width: 100, height: 100))
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) is unimplemented")
    }
    
    /** Random duration for the grow or shrink/fade animation. */
    private var resizeFadeDuration: Double
    {
        let kResizeFadeDuration: Double = 0.19
        let gamut: Double = 0.08
        let offset = randomOffset(withGamut: gamut)
        
        return kResizeFadeDuration + offset
    }

    /** Random duration for each of the parts of the size bounce animation when spawning. */
    private var bounceDuration: Double
    {
        let kBounceDuration: Double = 0.06
        let gamut: Double = 0.08
        let offset = randomOffset(withGamut: gamut)
        
        return kBounceDuration + offset
    }

    /** Random duration for a slide animation. */
    private var moveDuration: Double
    {
        let kMoveDuration: Double = 0.125
        let gamut: Double = 0.08
        let offset = randomOffset(withGamut: gamut)
        
        return kMoveDuration + offset
    }
    
    /** Create node at the given position, with animation. */
    func spawn(atPosition position: CGPoint)
    {
        let kInitialScale: CGFloat = 0.85
        let kBounceUpperScale: CGFloat = 1.1
        let kBounceLowerScale: CGFloat = 0.93
        let kUnityScale: CGFloat = 1.0
        
        let thisResizeDuration = self.resizeFadeDuration
        let thisBounceDuration = self.bounceDuration
        
        // Grow to full size
        self.setScale(kInitialScale)
        let grow = SKAction.scale(to: kUnityScale, duration: thisResizeDuration)
        
        // Do a little size bounce
        let scaleUp = SKAction.scale(to: kBounceUpperScale, duration: thisBounceDuration)
        let scaleDown = SKAction.scale(to: kBounceLowerScale, duration: thisBounceDuration)
        let scaleBackUp = SKAction.scale(to: kUnityScale, duration: thisBounceDuration)
        let bounce = SKAction.sequence([scaleUp, scaleDown, scaleBackUp])
        
        let spawn = SKAction.sequence([grow, bounce])
        
        self.position = position
        self.run(spawn)
    }
    
    /** Slide the node to the given position. */
    func move(toPosition destination: CGPoint)
    {
        self.move(toPosition: destination, duration: self.moveDuration)
    }
    
    /** Animate the node's dissappearance at the destination point. */
    func moveIntoCombination(atPosition destination: CGPoint)
    {
        let kFinalShrinkScale: CGFloat = 0.25
        
        let isChangingPosition = destination != self.position
        
        let thisResizeDuration = self.resizeFadeDuration
        
        let shrink = SKAction.scale(to: kFinalShrinkScale, duration: thisResizeDuration)
        let fade = SKAction.fadeOut(withDuration: thisResizeDuration)
        
        var fadeAndShrink = SKAction.group([shrink, fade])
        
        if isChangingPosition {
            
            let thisMoveDuration = self.moveDuration
            fadeAndShrink = SKAction.sequence([SKAction.wait(forDuration: thisMoveDuration), fadeAndShrink])
            self.move(toPosition: destination, duration: thisMoveDuration)
        }
        
        self.run(fadeAndShrink, completion: {
            
            // Next run loop
            DispatchQueue.main.async {
                self.removeFromParent()
            }
        })
    }
    
    /** Slide the node to the given position over the given duration. */
    private func move(toPosition destination: CGPoint, duration: TimeInterval)
    {
        let move = SKAction.move(to: destination, duration: duration)
        move.timingMode = .easeInEaseOut
        
        TFENode.movementDispatchGroup.enter()
        self.run(move, completion: {
            TFENode.movementDispatchGroup.leave()
        })
    }
}

//MARK: - Comparability
func ==(lhs: TFENode, rhs: TFENode) -> Bool
{
    return lhs.value == rhs.value
}

func <(lhs: TFENode, rhs: TFENode) -> Bool
{
    return lhs.value < rhs.value
}

extension TFENode : Comparable
{
    override var hashValue: Int { return super.hashValue << Int(self.value) }
}

//MARK: - Node movement tracking
extension TFENode
{
    /**
     * The TFENode class tracks whether any instances are running their "move" or
     * "spawn" animations in order to avoid overlapping creation and destruction
     * of moving nodes. `waitOnAllNodeMovement()` blocks when called until all
     * nodes have completed their animations.
     */
    class func waitOnAllNodeMovement()
    {
        _ = self.movementDispatchGroup.wait(timeout: .distantFuture)
    }
    
    /**
     * The TFENode class tracks whether any instances are running their "move" or
     * "spawn" animations. `anyNodeMovementInProgress()` returns immediately with a
     * boolean indicating whether any animations are currently running.
     */
    class func anyNodeMovementInProgress() -> Bool
    {
        // Return immediately regardless, but signal whether it was because
        // the group is inactive or because of timeout.
        return .success != self.movementDispatchGroup.wait(timeout: .now())
    }
    
    /**
     * The TFENode class tracks whether any instances are running animations.
     * This dispatch group is the means. An instance enters the group before
     * starting its animation action, and leaves the group in the action's
     * completion.
     */
    fileprivate static let movementDispatchGroup = DispatchGroup()
}
